From 28974dba1ac5badfa5ceb1267b1ed394cfce6190 Mon Sep 17 00:00:00 2001
From: Md Sadre Alam <quic_mdalam@quicinc.com>
Date: Tue, 28 Nov 2023 12:47:32 +0530
Subject: [PATCH 479/500] crypto: qce - Fix sleftest and tcrypt test issue

Fix self test and tcrypt test for all the crypto
algorithm when cmd descriptor support enabled.

Once command descriptor enabled, then all the
crypto register read/write will happen via crypto
bam. So check the status of a transfer before
terminating dma else encryption/decryption will fail
randomly due to bad status update in status register.

error:

[   99.984898] alg: skcipher: cbc-des-qce encryption failed on test vector 4; expected_error=0, actual_error=-6, cfg="uneven misaligned splits, may sleep"
[   99.985760] alg: self-tests for cbc(des) using cbc(des) failed (rc=-6)
[   99.985762] ------------[ cut here ]------------
[  100.005630] alg: self-tests for cbc(des) using cbc(des) failed (rc=-6)
[  100.005651] WARNING: CPU: 3 PID: 5006 at crypto/testmgr.c:5929 alg_test+0x360/0x3bc
[  100.298503] CPU: 3 PID: 5006 Comm: insmod Tainted: G        W          6.1.31 #0
[  100.320733] Hardware name: Qualcomm Technologies, Inc. IPQ9574/RDP418/AP-AL02-C1 (DT)
[  100.328198] pstate: 60400005 (nZCv daif +PAN -UAO -TCO -DIT -SSBS BTYPE=--)
[  100.335923] pc : alg_test+0x360/0x3bc
[  100.342691] lr : alg_test+0x360/0x3bc
[  100.346510] sp : ffffff801141b9d0
[  100.350156] x29: ffffff801141b9d0 x28: 000000000000000c x27: 0000000000000001
[  100.353457] x26: 0000000000000000 x25: ffffffc008bf2428 x24: 0000000000000000
[  100.360575] x23: 0000000000000000 x22: ffffffc0014af5eb x21: 0000000000000027
[  100.367694] x20: ffffffc0014af5eb x19: 00000000fffffffa x18: 0000000000000000
[  100.374812] x17: 0000000000000000 x16: 0000000000000000 x15: 000000000000002d
[  100.381929] x14: 0000000000000001 x13: 0000000000000001 x12: 0000000000000000
[  100.389049] x11: 0000000000000000 x10: 0000000000000960 x9 : ffffff801141b870
[  100.396167] x8 : ffffff800fb309c0 x7 : ffffff803ff846c0 x6 : 0000000000000000
[  100.403284] x5 : 00000000108b4000 x4 : 0000000000000003 x3 : ffffff800f916dc8
[  100.410402] x2 : 0000000000000000 x1 : 0000000000000000 x0 : ffffff800fb30000
[  100.417522] Call trace:
[  100.424634]  alg_test+0x360/0x3bc
[  100.426891]  tcrypt_test+0x44/0x50 [tcrypt]
[  100.430365]  do_test+0x138/0x3de4 [tcrypt]
[  100.434357]  tcrypt_mod_init+0x54/0x1000 [tcrypt]
[  100.438525]  do_one_initcall+0x68/0x188
[  100.443298]  do_init_module+0x4c/0x1b4
[  100.446944]  load_module+0x1678/0x1704
[  100.450763]  __do_sys_init_module+0x208/0x234
[  100.454497]  __arm64_sys_init_module+0x14/0x1c
[  100.458925]  el0_svc_common.constprop.0+0xc0/0x130
[  100.463264]  do_el0_svc+0x18/0x20
[  100.468037]  el0_svc+0x14/0x3c
[  100.471423]  el0t_64_sync_handler+0xa8/0x134
[  100.474375]  el0t_64_sync+0x14c/0x150
[  100.478802] ---[ end trace 0000000000000000 ]---
[  100.482383] tcrypt: testing ctr(des)
[  100.487299] tcrypt: one or more tests failed!
failed to insert tcrypt.ko

Change-Id: I16ae35bd4564e803b31554388d963b20ff701267
Signed-off-by: Md Sadre Alam <quic_mdalam@quicinc.com>
---
 drivers/crypto/qce/aead.c     | 11 +++++++----
 drivers/crypto/qce/common.c   |  7 ++-----
 drivers/crypto/qce/sha.c      | 11 +++++++----
 drivers/crypto/qce/skcipher.c | 11 +++++++----
 4 files changed, 23 insertions(+), 17 deletions(-)

diff --git a/drivers/crypto/qce/aead.c b/drivers/crypto/qce/aead.c
index bad6ebe22bda..3fd6856a828e 100644
--- a/drivers/crypto/qce/aead.c
+++ b/drivers/crypto/qce/aead.c
@@ -42,6 +42,13 @@ static void qce_aead_done(void *data)
 	dir_src = diff_dst ? DMA_TO_DEVICE : DMA_BIDIRECTIONAL;
 	dir_dst = diff_dst ? DMA_FROM_DEVICE : DMA_BIDIRECTIONAL;
 
+	error = qce_check_status(qce, &status);
+	if (error < 0 && (error != -EBADMSG))
+		dev_err(qce->dev, "aead operation error (%x)\n", status);
+
+	if (qce->qce_cmd_desc_enable)
+		qce_unlock_reg_dma(qce);
+
 	error = qce_dma_terminate_all(&qce->dma);
 	if (error)
 		dev_dbg(qce->dev, "aead dma termination error (%d)\n",
@@ -77,10 +84,6 @@ static void qce_aead_done(void *data)
 		sg_free_table(&rctx->dst_tbl);
 	}
 
-	error = qce_check_status(qce, &status);
-	if (error < 0 && (error != -EBADMSG))
-		dev_err(qce->dev, "aead operation error (%x)\n", status);
-
 	if (IS_ENCRYPT(rctx->flags)) {
 		totallen = req->cryptlen + req->assoclen;
 		if (IS_CCM(rctx->flags))
diff --git a/drivers/crypto/qce/common.c b/drivers/crypto/qce/common.c
index 8c02aa1064cf..c0c4c6776e9e 100644
--- a/drivers/crypto/qce/common.c
+++ b/drivers/crypto/qce/common.c
@@ -636,7 +636,7 @@ int qce_read_dma_get_lock(struct qce_device *qce)
 	u32 val = 0;
 
 	qce_clear_bam_transaction(qce);
-	qce_read_reg_dma(qce, REG_STATUS, &val, 1);
+	qce_read_reg_dma(qce, REG_STATUS2, &val, 1);
 
 	ret = qce_submit_cmd_desc(qce, QCE_DMA_DESC_FLAG_LOCK);
 	if (ret) {
@@ -653,7 +653,7 @@ int qce_unlock_reg_dma(struct qce_device *qce)
 	u32 val = 0;
 
 	qce_clear_bam_transaction(qce);
-	qce_read_reg_dma(qce, REG_STATUS, &val, 1);
+	qce_read_reg_dma(qce, REG_STATUS2, &val, 1);
 
 	ret = qce_submit_cmd_desc(qce, QCE_DMA_DESC_FLAG_UNLOCK);
 	if (ret) {
@@ -983,9 +983,6 @@ int qce_check_status(struct qce_device *qce, u32 *status)
 
 	*status = qce_read(qce, REG_STATUS);
 
-	/* Unlock the crypto pipe here */
-	if (qce->qce_cmd_desc_enable)
-		qce_unlock_reg_dma(qce);
 	/*
 	 * Don't use result dump status. The operation may not be complete.
 	 * Instead, use the status we just read from device. In case, we need to
diff --git a/drivers/crypto/qce/sha.c b/drivers/crypto/qce/sha.c
index 8b4a5d7968ae..55c4ca1f7858 100644
--- a/drivers/crypto/qce/sha.c
+++ b/drivers/crypto/qce/sha.c
@@ -47,6 +47,13 @@ static void qce_ahash_done(void *data)
 	int error;
 	u32 status;
 
+	error = qce_check_status(qce, &status);
+	if (error < 0)
+		dev_dbg(qce->dev, "ahash operation error (%x)\n", status);
+
+	if (qce->qce_cmd_desc_enable)
+		qce_unlock_reg_dma(qce);
+
 	error = qce_dma_terminate_all(&qce->dma);
 	if (error)
 		dev_dbg(qce->dev, "ahash dma termination error (%d)\n", error);
@@ -74,10 +81,6 @@ static void qce_ahash_done(void *data)
 				DMA_MEM_TO_DEV);
 	}
 
-	error = qce_check_status(qce, &status);
-	if (error < 0)
-		dev_dbg(qce->dev, "ahash operation error (%x)\n", status);
-
 	req->src = rctx->src_orig;
 	req->nbytes = rctx->nbytes_orig;
 	rctx->last_blk = false;
diff --git a/drivers/crypto/qce/skcipher.c b/drivers/crypto/qce/skcipher.c
index 2201d3d5883c..0fa64f63c165 100644
--- a/drivers/crypto/qce/skcipher.c
+++ b/drivers/crypto/qce/skcipher.c
@@ -51,6 +51,13 @@ static void qce_skcipher_done(void *data)
 	dir_src = diff_dst ? DMA_TO_DEVICE : DMA_BIDIRECTIONAL;
 	dir_dst = diff_dst ? DMA_FROM_DEVICE : DMA_BIDIRECTIONAL;
 
+	error = qce_check_status(qce, &status);
+	if (error < 0)
+		dev_dbg(qce->dev, "skcipher operation error (%x)\n", status);
+
+	if (qce->qce_cmd_desc_enable)
+		qce_unlock_reg_dma(qce);
+
 	error = qce_dma_terminate_all(&qce->dma);
 	if (error)
 		dev_dbg(qce->dev, "skcipher dma termination error (%d)\n",
@@ -75,10 +82,6 @@ static void qce_skcipher_done(void *data)
 				DMA_MEM_TO_DEV);
 	}
 
-	error = qce_check_status(qce, &status);
-	if (error < 0)
-		dev_dbg(qce->dev, "skcipher operation error (%x)\n", status);
-
 	memcpy(rctx->iv, result_buf->encr_cntr_iv, rctx->ivsize);
 	qce->async_req_done(tmpl->qce, error);
 }
-- 
2.34.1

